Feat/undo actions

sherinRoy opened the pull request.

kandarp-patel51 commented on apps/okreads-e2e/src/support/commands.ts:
```diff
@@ -8,6 +8,11 @@ declare global {
   namespace Cypress {
     interface Chainable {
       startAt: typeof startAt;
+      searchBooks: typeof searchBooks;
```
I don't see these changes are used anywhere.
Make sure that you are not committing any unwanted changes

kandarp-patel51 replied:
Pls remove these file changes

kandarp-patel51 commented on libs/books/data-access/src/lib/+state/reading-list.effects.ts:
```diff
@@ -54,9 +56,59 @@ export class ReadingListEffects implements OnInitEffects {
+
+  openSnackBar(item: ReadingListItem | Book, message: string, isAdded: boolean): void {
+    this.snackBar.open(message, 'undo', {
+      duration: 3000
```
Create constant instead of using hardcoded value here

kandarp-patel51 replied:
Pls make this change

kandarp-patel51 reviewed (changes requested):
Pls take a look at review comments

sherinRoy commented:
Undo Add and Remove from reading list

- Implemented undo functionality on addition/removal of book.
- On successful addition to reading list or on successful removal from reading list, snack bar will open with undo button.
- Added test cases to ensure the functionality.

Screen shots

- Lint
  
<img width="1512" alt="Screenshot 2023-11-28 at 11 40 09 PM" src="https://github.com/sherinRoy/web-ui-developer-puzzle/assets/148328772/14702209-7c0d-4447-8158-9565ba1c0d7e">

- Unit test
 
<img width="1512" alt="Screenshot 2023-11-28 at 11 40 56 PM" src="https://github.com/sherinRoy/web-ui-developer-puzzle/assets/148328772/4601d453-a625-4d6b-8795-7c0e452bdfb7">

- e2e test
<img width="1512" alt="Screenshot 2023-11-28 at 11 52 20 PM" src="https://github.com/sherinRoy/web-ui-developer-puzzle/assets/148328772/39936898-2a0b-4a15-8744-a0fb6786b9b3">


https://github.com/sherinRoy/web-ui-developer-puzzle/assets/148328772/4873ea4f-0d73-4abf-900e-a5a01722fd72


https://github.com/sherinRoy/web-ui-developer-puzzle/assets/148328772/0a9e2209-cbae-4e5a-b27f-b4e03644f6c9

kandarp-patel51 commented on apps/okreads-e2e/src/integration/reading-list.spec.ts:
```diff
@@ -1,14 +1,58 @@
   });
 
-  it('Then: I should see my reading list', () => {
+  it('Then: I should see my reading list', async () => {
```
Do we need async here?

kandarp-patel51 commented on apps/okreads-e2e/src/support/commands.ts:
```diff
@@ -5,16 +5,17 @@
 // ***********************************************
 
```
Please revert this file changes.

kandarp-patel51 commented on libs/books/feature/src/lib/reading-list/reading-list.component.html:
```diff
@@ -21,6 +21,7 @@
         color="warn"
         [attr.aria-label]="'Remove ' + book.title + ' from reading list'"
         (click)="removeFromReadingList(book)"
+        data-testing="remove-item"
```
Suggestion: We can take it like : ```data-testing="remove-book-from-list"```

kandarp-patel51 commented on libs/books/feature/src/lib/book-search/book-search.component.html:
```diff
@@ -26,7 +26,13 @@
-            <button mat-flat-button color="primary" (click)="addBookToReadingList(book)" [disabled]="book.isAdded">
+            <button
+              mat-flat-button
+              data-testing="add-item"
```
Suggestion: We can take it like : ```data-testing="add-book-to-list"```

kandarp-patel51 reviewed (changes requested):
Pls have a look at review comment

kandarp-patel51 commented on libs/books/feature/src/lib/book-search/book-search.component.html:
```diff
@@ -26,7 +26,13 @@
           <p [innerHTML]="book.description"></p>
           <div>
-            <button mat-flat-button color="primary" (click)="addBookToReadingList(book)" [disabled]="book.isAdded">
+            <button
```
Avoid such changes

kandarp-patel51 reviewed (changes requested):
Take a look at review comments
